File: 647-palindromic-substrings/palindromic-substrings.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        n = len(s)
        sol = n

        #odd
        dp = [True] * n
        for i in range(2,n,2):
            for j in range(i, n):
                if s[j] == s[j-i] and dp[j-i+1]:
                    dp[j-i] = True
                    sol+= 1
                else:
                    dp[j-i] = False
        
        #Even
        dp = [True] * n

        for i in range(1,n,2):
            for j in range(i, n):
                if s[j] == s[j-i] and dp[j-i+1]:
                    dp[j-i] = True
                    sol+= 1
                else:
                    dp[j-i] = False
        
        return sol

File: 647-palindromic-substrings/palindromic-substrings.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int n = s.size();

        vector<vector<bool>> dp(n, vector<bool> (n, true));
        int ans = n;

        for(int i = n-2; i >= 0; i--){
            for(int j = n-1; j>i ; j--){
                dp[i][j] = s[i] == s[j] && dp[i+1][j-1];
                if(dp[i][j]){
                    ans += 1;
                }
            }
        }
        return ans;
    }
};